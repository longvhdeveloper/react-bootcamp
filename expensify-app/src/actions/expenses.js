import uuid from 'uuid'

// add expense
export const addExpense = ({
  description = '',
  notes = '',
  amount = 0,
  createAt = 0,
} = {}) => ({
  type: 'ADD_EXPENSE',
  expense: {
    id: uuid(),
    description,
    notes,
    amount,
    createAt,
  },
})

// remove expense
export const removeExpense = ({ id = '' } = {}) => {
  return {
    type: 'REMOVE_EXPENESE',
    id,
  }
}

// edit expense
export const editExpense = (id, updates = {}) => {
  return {
    type: 'EDIT_EXPENSE',
    id,
    updates,
  }
}
