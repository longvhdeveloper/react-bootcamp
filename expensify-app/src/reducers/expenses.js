// expense reducer
const expensesReducerDefaultState = []

export default (state = expensesReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_EXPENSE':
      return [...state, action.expense]
    case 'REMOVE_EXPENESE':
      return state.filter((expense) => expense.id != action.id)
    case 'EDIT_EXPENSE':
      return state.map((expense) => {
        if (expense.id === action.id) {
          return {
            ...expense,
            ...action.updates,
          }
        } else {
          return expense
        }
      })
    default:
      return state
  }
}
