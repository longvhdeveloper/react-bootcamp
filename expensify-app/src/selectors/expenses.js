import moment from 'moment'

// get visible expenses
export default (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses
    .filter((expense) => {
      const createAtMoment = moment(expense.createAt)
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createAtMoment, 'date')
        : true

      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createAtMoment, 'date')
        : true
      const textMacth = expense.description
        .toLowerCase()
        .includes(text.toLowerCase())

      return startDateMatch && endDateMatch && textMacth
    })
    .sort((a, b) => {
      if (sortBy === 'date') {
        return a.createAt < b.createAt ? 1 : -1
      }

      if (sortBy === 'amount') {
        return a.amount < b.amount ? 1 : -1
      }
    })
}
