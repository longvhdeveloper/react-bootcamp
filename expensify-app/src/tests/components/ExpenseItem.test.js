import React from 'react'
import { shallow } from 'enzyme'
import ExpneseItem, { ExpenseItem } from '../../components/ExpenseItem'
import expenses from '../fixtures/expenses'
import { expect } from '@jest/globals'

test('should render ExpenseItem with expense', () => {
  const wrapper = shallow(<ExpneseItem expense={expenses[0]} />)
  expect(wrapper).toMatchSnapshot()
})
