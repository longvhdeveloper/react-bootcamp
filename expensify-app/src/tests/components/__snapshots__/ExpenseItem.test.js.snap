// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseItem with expense 1`] = `
<tr>
  <td>
    Dum
  </td>
  <td />
  <td>
    195
  </td>
  <td>
    0
  </td>
  <td>
    <[object Object]
      to="/edit/1"
    >
      Edit
    </[object Object]>
  </td>
</tr>
`;
